src/NuGet.Protocol.Core.v3/ServiceIndexResourceV3Provider.cs

```diff
@@ -56,29 +59,61 @@ public ServiceIndexResourceV3Provider()
                 {
-                    var messageHandlerResource = await source.GetResourceAsync<HttpHandlerResource>(token);
-
-                    var client = new DataClient(messageHandlerResource);
-
-                    JObject json;
+                    var uri = new Uri(url);
+                    ICredentials credentials = CredentialStore.Instance.GetCredentials(uri);
+                    JObject json = null;
 
-                    try
-                    {
-                        json = await client.GetJObjectAsync(new Uri(url), token);
-                    }
-                    catch (JsonReaderException)
+                    while (true)
                     {
-                        var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
-                        _cache.AddOrUpdate(url, entry, (key, value) => entry);
+                        var messageHandlerResource = await source.GetResourceAsync<HttpHandlerResource>(token);
+                        if (credentials != null)
+                        {
+                            messageHandlerResource.ClientHandler.Credentials = credentials;
+                        }
 
-                        return new Tuple<bool, INuGetResource>(false, null);
-                    }
-                    catch (HttpRequestException)
-                    {
-                        var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
-                        _cache.AddOrUpdate(url, entry, (key, value) => entry);
+                        try
+                        {
+                            using (var client = new DataClient(messageHandlerResource))
+                            {
+                                json = await client.GetJObjectAsync(uri, token);
+                                if (HttpHandlerResourceV3.CredentialsSuccessfullyUsed != null && credentials != null)
+                                {
+                                    HttpHandlerResourceV3.CredentialsSuccessfullyUsed(uri, credentials);
+                                }
+
+                                break;
+                            }
+                        }
+                        catch (JsonReaderException)
+                        {
+                            var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
+                            _cache.AddOrUpdate(url, entry, (key, value) => entry);
 
-                        return new Tuple<bool, INuGetResource>(false, null);
+                            return new Tuple<bool, INuGetResource>(false, null);
+                        }
+                        catch (HttpRequestException)
+                        {
+                            var entry = new ServiceIndexCacheInfo { CachedTime = utcNow };
+                            _cache.AddOrUpdate(url, entry, (key, value) => entry);
+
+                            return new Tuple<bool, INuGetResource>(false, null);
```

Comment:
This will lead to a large number of requests for index.json when a v3 source is unreachable or the user is offline. Please verify that this doesn't cause hangs and that we really want to keep retrying this.